### PR TITLE
Start of rio-toa

### PR DESCRIPTION
**Pull request by dnomadb**
@celoyd I have a basic, non rescaled implementation of `rio toa radiance` based on http://landsat.usgs.gov/Landsat8_Using_Product.php going here. Right now, it only outputs an unscaled `float32` raster.

Todo:
- [x] Documentation - inline + w/ readme
- [x] Check math
  ~~\- [ ] Handle rescaling~~
- [x] Write `rio toa reflectance`
  ~~\- [ ] Write tests~~

cc @perrygeo @virginiayung

### CODE REVIEW

**Comment by perrygeo** [comment]
👍 awesome idea and implementation. I like how you've split the functions up - that's a great pattern to have the core `radiance` function take only arrays and raw parameters.

A few comments, none of which should hold up merging this:
- why not float64 for the maths?
- a function to scale to uint8 would be nice eventually
- more of a rio mucho question but I find multiprocessing difficult to work with while debugging; it screws up code coverage and profiling, often gives opaque errors and prevents the use of pdb. I started using [this pattern](https://github.com/mapbox/rio-color/blob/master/rio_color/scripts/cli.py#L93-L110) in rio-color which calls the function "normally", bypassing rio mucho if workers is 1.
- having a CLI to go from `1 band -> 1 band tif` is nice and clean. But there might room for a variant or a higher-level command that takes `3 tifs -> 1 rgb tif` (only writes to disk once)

**Comment by dnomadb** [comment]
> why not float64 for the maths?

Y this was sort of a seat-of-the-pants decision. I arbitrarily picked `float32`, mainly b/c I seemed to recall more difficulty in testing `float64` (precision errors? idk), and `float32` seemed ok precision wise. 👍 to changing it if you think it is better.

> a function to scale to uint8 would be nice eventually

Totally, this is what I was alluding to with rescaling. @celoyd and I were talking about this, and we'll prob want to clip off some global (to the entire set of scenes) variable? Maybe we have a `--clip (<b>, <min> <max>)` option?

> more of a rio mucho question but I find multiprocessing difficult to work with while debugging;

I agree. Do you think this is worthwhile to solve in `rio-mucho` ? IE, if `jobs == 1` / `debug == True`, mock the functionality and process normally?

> But there might room for a variant or a higher-level command that takes 3 tifs -> 1 rgb tif

Yes, I had this thought as well - would make this a stacker as well. We'd have to work out how to work w/in the band opts logic (which band is which, how to pass clip args if there are any, etc)

**Comment by perrygeo** [comment]
float64 is generally better with precision but in this case it probably doesn't matter much. I just always use it by default, wondered if float32 was an intentional optimization for something.

Rio mucho thing ticketed: https://github.com/mapbox/rio-mucho/issues/9

CLI enhancements above all sounds great but need not hold up a 0.1 release.

**Comment by celoyd** [comment]
I think `float32` makes sense here because:
- the values are actually 12-bit,
- only two arithmetic operations happen, an addition and a multiplication (so the opportunity for drift is very small), and
- for this application, we’re not concerned about bit-perfect accuracy.

If we want a “use `float64` everywhere” blanket rule for simplicity, I’m fine with that. Just saying that I think this is an appropriate place for the small optimization of `float32`.

File: rio_toa/reflectance.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+    """
+
+    rf = (MR * img.astype(np.float32)) + AR) / sin(E)
+    rf[img == src_nodata] = 0.0
+
+    return rf
+
+
+def reflectance_worker(data, window, ij, g_args):
```
Comment (dnomadb) [inline rio_toa/reflectance.py]: @virginiayung let's make these methods private by default (prefixed w/ an underscore, ie `_reflectance_worker`

**Comment by dnomadb** [comment]
@virginiayung to merge this, lets:
- Wait to add any more tests. We can both branch from master and work on tests for each function there
- Wait to add any more CLI complexity as of yet
- Update the readme to reflect what the intent and basic functionality of this should be

**Comment by celoyd** [comment]
L:+1:TM!

**Pull request merged**